add(mempool): Verify transactions with unmined inputs in the mempool

arya2 opened the pull request:
## Motivation

We want to verify transactions in the mempool that spend the transparent outputs of other transactions in the mempool.

Closes #8777.

## Solution

- Adds a handle to the mempool service in the transaction verifier
- Adds `AwaitOutput` request in the mempool service
- Updates the `spent_utxos()` method in the transaction verifier to try querying the mempool for spent outpoints before returning a `TransparentInputNotFound` error.
- Polls the mempool service from the transaction verifier after verifying a mempool transaction that creates new UTXOs (in case they're spent by other mempool transactions that are still being verified)
- Updates `getblocktemplate` method and ZIP-317 transaction selection to include transactions that spent the outputs of other transactions in the mempool when all of their dependencies have been selected

### Tests

Added tests to check that:
- Inserting a transaction also updates the transaction dependencies tracked in the verified set
- The `Storage.remove()` method works as expected (the rest of `evict_one()` should be covered already)
- The mempool responds to `AwaitOutput` requests when the output for the queried outpoint is:
  - Added to the mempool
  - Already available in the mempool
- The transaction verifier:
  - Calls the mempool with `AwaitOutput` requests as expected and processes `UnspentOutput` responses correctly
  - Polls the mempool after verifying a mempool transaction that creates new transparent outputs
  - Includes all of the expected `spent_mempool_outpoints` in its response
- Long poll ids remaining consistent as long the mempool contents are unchanged (and everything else is unchanged)
- ZIP-317 mempool transaction selection excludes transactions that have dependencies unless their dependencies have already been selected

If needed, this PR could also add tests for:
- ZIP-317 mempool transaction selection always returns the right order of transactions (this could be unnecessary since they’re sorted afterwards anyway but it could be a good sanity check)
- The `getblocktemplate` method returns transactions with dependencies in the right order when the transaction hashes would have them sorted in the wrong order
- Ideally at least a basic acceptance test that checks that it all works via RPCs

## Follow Up Work

The mempool's `Downloads` sends a result to the response channel too early, it should wait until the transaction is being inserted into the mempool's verified set.

### PR Author's Checklist

<!-- If you are the author of the PR, check the boxes below before making the PR
ready for review. -->

- [x] The PR name will make sense to users.
- [ ] The PR provides a CHANGELOG summary.
- [x] The solution is tested.
- [x] The documentation is up to date.
- [x] The PR has a priority label.

### PR Reviewer's Checklist

<!-- If you are a reviewer of the PR, check the boxes below before approving it. -->

- [ ] The PR Author's checklist is complete.
- [ ] The PR resolves the issue.

arya2 committed:
Adds a parameter to `zebra_consensus::router::init()` for accepting a mempool setup argument, adds and uses an `init_test()` fn for passing a closed channel receiver in tests where no mempool service is needed in the transaction verifier.

arya2 committed:
Adds a `mempool` argument to the transaction::Verifier constructor (and a `new_for_tests()` constructor for convenience)

arya2 committed:
Removes `Clone` impl on `transaction::Verifier` to add mempool oneshot receiver, updates tests.

arya2 committed:
Adds TODOs

arya2 committed:
updates transaction verifier's poll_ready() method to setup the mempool service handle.

arya2 committed:
Updates VerifiedSet struct used in mempool storage

arya2 committed:
Updates mempool service and its `Storage` to use the updated `VerifiedSet` `transactions()` return type.

arya2 committed:
updates `created_outputs` when inserting or removing a transaction from the mempool's verified set

arya2 committed:
Adds a TODO, updates field docs

arya2 committed:
Updates `spent_utxos()` to query the mempool for unspent outputs

arya2 committed:
Adds `spent_mempool_outpoints` as a field on tx verifier mempool response

arya2 committed:
Updates mempool `Downloads` to return the spent_mempool_outpoints from the tx verifier response

arya2 committed:
Updates `Storage.insert()` to accept a list of spent mempool transaction outputs

arya2 committed:
Adds transaction dependencies when inserting a tx in `VerifiedSet`

arya2 committed:
polls mempool svc from tx verifier when a mempool tx that creates transparent outputs has been verified.

adds a TODO for adding a `pending_outputs` field to the mempool Storage

arya2 committed:
Adds `pending_outputs` field on mempool Storage and responds to pending outputs requests when inserting new transactions into the mempool's verified set

arya2 committed:
replaces `UnminedTxId` type with `transaction::Hash` in mempool's verified set

arya2 committed:
prune pending outputs when rejecting and removing same effects.

arya2 committed:
Remove dependent transactions from verified set when removing a tx

arya2 committed:
updates tests

arya2 committed:
appeases clippy.

arya2 committed:
removes unused `len()` method

arya2 committed:
fixes doc links

arya2 committed:
Adds transaction dependencies to the `FullTransactions` response, let the caller handle it (required to avoid moving zip317 tx selection code to mempool)

arya2 committed:
updates block template construction to avoid including transactions unless their dependencies have already been added.

arya2 committed:
updates tests

arya2 committed:
Replaces placeholder setup channel with one that sends the mempool svc to the tx verifier, adds a timeout layer, adds a TODO about a concurrency bug

arya2 committed:
Use a single query to check for unspent outputs in the mempool

arya2 committed:
Updates `getblocktemplate` method to consider dependencies when sorting transactions for the final template

arya2 committed:
fixes clippy lints, removes unnecessary Option in UnspentOutput response variant

arya2 committed:
renames type alias and method, adds a TODO to use iteration instead of recursion

arya2 committed:
Adds mempool_removes_dependent_transactions() test

arya2 committed:
Updates Storage and VerifiedSet clear() methods to clear pending_outputs, created_outputs, and transaction_dependencies, adds TODO to use iteration instead of recursion.

arya2 committed:
removes outdated TODO

arya2 committed:
Adds a TODO for reporting queued transaction verification results from the mempool from the poll_ready() method

arya2 committed:
Adds `mempool_responds_to_await_output` test

arya2 committed:
updates mempool_responds_to_await_output test

arya2 committed:
Uses iteration instead of recursion in verified set's remove() method and zip317 mod's dependencies_depth() method

arya2 committed:
Adds a mempool_request_with_mempool_output_is_accepted test for the transaction verifier

arya2 committed:
Moves delay duration before polling the mempool to a constant, uses a shorter timeout for mempool output lookups, adds a `poll_count` to MockService, and updates `mempool_request_with_unmined_output_spends_is_accepted` to check that the transaction verifier polls the mempool after verifying a mempool transaction with transparent outputs

arya2 committed:
adds long_poll_input_mempool_tx_ids_are_sorted test

arya2 committed:
Adds a `excludes_tx_with_unselected_dependencies` test

arya2 committed:
Updates a TODO

arya2 committed:
moves `TransactionDependencies` struct to `zebra-node-services`

arya2 committed:
Updates `FullTransactions` response variant's `transaction_dependencies` type

arya2 committed:
updates zip317 transaction selection for block templates to include dependent transactions

arya2 committed:
Moves and refactors zip317 tx selection test to its own module, adds an `unmined_transactions_in_blocks()` method on network

arya2 committed:
Removes `unmined_transactions_in_blocks()`  test utility fn from mempool Storage test module and replaces calls to it with calls to the new test method on Network

arya2 committed:
Fixes spelling mistake

arya2 committed:
Adds `includes_tx_with_selected_dependencies` test

arya2 commented on zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs:
```diff
@@ -188,23 +255,90 @@ fn checked_add_transaction_weighted_random(
+                let mempool_tx_deps = mempool_tx_deps.dependencies();
+                let candidate_tx_deps = mempool_tx_deps.get(dependent_candidate_tx_id);
+
+                if has_direct_dependencies(candidate_tx_deps, selected_txs) {
```
~If anyone has any ideas on how to address this, there's an issue here that I'm not sure how to fix this yet where tx3 won't be selected if tx1 is independent, and tx2 depends on tx1, but tx3 depends on both tx1 and tx2, because it's checked before tx2 is selected.~

~I thought about storing those in another collection to try selecting them after the rest of the dependent txs, but that still won't work if there's a tx4 that depends on tx1 and tx3.~

Update: There's a simple fix here. I think it may have been a non-issue to begin with, but the new code looks simpler and more efficient anyways.

arya2 committed:
fixes zip317 block construction issue

arya2 committed:
Fixes vectors test

arya2 commented on zebra-node-services/src/mempool.rs:
```diff
@@ -14,6 +17,12 @@ use crate::BoxError;
+mod transaction_dependencies;
+
+#[cfg(feature = "getblocktemplate-rpcs")]
+pub use transaction_dependencies::TransactionDependencies;
```
```suggestion
mod transaction_dependencies;

pub use transaction_dependencies::TransactionDependencies;
```

arya2 committed:
Update zebra-node-services/src/mempool.rs

arya2 commented on zebra-consensus/src/transaction.rs:
```diff
@@ -473,7 +547,17 @@ where
+                                mempool.ready().await.expect("mempool poll_ready() method should not return an error");
+                            });
+                        }
+                    }
```
This should be moved to the mempool's `Downloads` component, see [this comment](https://github.com/ZcashFoundation/zebra/pull/8901/files#r1794389135).

The changes in this PR seem excessively substantial already, so I'll do this in a follow up PR.

conradoplg commented on zebrad/src/components/mempool/storage.rs:
```diff
@@ -116,12 +127,15 @@ pub struct Storage {
     ///
     /// Only transactions with the exact [`UnminedTxId`] are invalid.
-    tip_rejected_exact: HashMap<UnminedTxId, ExactTipRejectionError>,
+    tip_rejected_exact: HashMap<transaction::Hash, ExactTipRejectionError>,
```
What is the reason for this change? The semantics of `tip_rejected_exact` is for it to incude transactions that have failed to validate e.g. due to an invalid signature. If this is changed to a Hash, then a transaction with e.g. the same data but a valid signature would be rejected when it shouldn't.

arya2 replied:
> What is the reason for this change?

I think the broader change to mined ids was to avoid iterating through the verified set to find the auth digest when adding transaction dependencies (since `OutPoint` only includes the mined id).

The change to `tip_rejected_exact` was a mistake, it would work fine for the `sendrawtransaction` RPC, but I see it would incorrectly reject gossiped transaction ids when it should try to download and verify them. 

Restored its type to the full `UnminedTxId` in https://github.com/ZcashFoundation/zebra/pull/8857/commits/ce6d169a9a0a1848b1d50a9eb29b7710f48043da.

conradoplg commented on zebrad/src/components/mempool/storage.rs:
```diff
@@ -180,11 +195,16 @@ impl Storage {
+    pub fn insert(
+        &mut self,
+        tx: VerifiedUnminedTx,
+        spent_mempool_outpoints: Vec<transparent::OutPoint>,
```
Please document the new argument

arya2 replied:
Thank you for catching that, documented in https://github.com/ZcashFoundation/zebra/pull/8857/commits/cc76b40885215fa1903ad26349583ac863f22f54.

conradoplg reviewed (commented):
I'm still reviewing but I added a comment

arya2 committed:
restores `tip_rejected_exact` type

arya2 committed:
updates affected tests

arya2 committed:
Documents the new argument in `Storage::insert()`, updates outdated comment

arya2 commented on zebrad/src/components/mempool/storage/verified_set.rs:
```diff
@@ -126,15 +143,35 @@ impl VerifiedSet {
 
-        self.cache_outputs_from(&transaction.transaction.transaction);
+        // This likely only needs to check that the transaction hash of the outpoint is still in the mempool,
+        // bu it's likely rare that a transaction spends multiple transparent outputs of
```
```suggestion
        // but it's likely rare that a transaction spends multiple transparent outputs of
```

arya2 committed:
Update zebrad/src/components/mempool/storage/verified_set.rs

upbqdn committed:
Merge branch 'main' into verify-orphaned-mempool-txs

arya2 committed:
fixes potential issue with calling buffered mempool's poll_ready() method without calling it.

arya2 committed:
Avoids removing dependent transactions of transactions that have been mined onto the best chain.

conradoplg reviewed (dismissed):
This is a complex PR (not your fault, it's a complex issue). I've been reviewing it for a while but it feels I could spend ages reviewing this to be 100% sure. I'm OK with approving it but do you think we should ask for someone else to take a look?

Also, did you manage to test it for the use case we wish to address (i.e. sending to tex addresses?)

arya2 commented:
> This is a complex PR (not your fault, it's a complex issue). I've been reviewing it for a while but it feels I could spend ages reviewing this to be 100% sure. I'm OK with approving it but do you think we should ask for someone else to take a look?

In hindsight, it probably should've been split into multiple PRs. A second review would be ideal since it involves concurrency and consensus changes, though it has also gone through a few rounds of self-review.

I think a second review of just the changes to ZIP 317 block production would be okay.

> Also, did you manage to test it for the use case we wish to address (i.e. sending to tex addresses?)

I haven't, and there's no acceptance test that tries to submit an orphaned transaction, I'll try doing that now.

arya2 committed:
Merge remote-tracking branch 'origin/main' into verify-orphaned-mempool-txs

upbqdn commented:
I'm also still reviewing. A bunch of other things took precedence in the meantime, unfortunately.

upbqdn commented on zebra-consensus/src/transaction.rs:
```diff
@@ -52,31 +60,79 @@ mod tests;
 {
     /// Create a new transaction verifier.
-    pub fn new(network: &Network, state: ZS) -> Self {
+    pub fn new(network: &Network, state: ZS, mempool_setup_rx: oneshot::Receiver<Mempool>) -> Self {
```
Can we not go without a channel?

arya2 replied:
It needs some way to get it concurrently since the mempool also has a handle to the transaction verifier. It doesn't have to be a channel, but a channel seemed like the simplest option and consistent with how the inbound service is set up.

upbqdn commented on zebra-consensus/src/transaction.rs:
```diff
@@ -523,19 +617,25 @@ where
     /// in the same order as the matching inputs in the transaction.
     async fn spent_utxos(
```
We need to update the fn docs.

arya2 replied:
Updated in https://github.com/ZcashFoundation/zebra/pull/8857/commits/257567b559dba0359519d61684e80daac52cc434.

upbqdn replied:
I was referring to:

- The description of `known_utxos`, since the fn doesn't take that anymore. We could move that comment somewhere else.
- The returned values, since the fn doesn't return a tuple anymore.

upbqdn commented on zebra-consensus/src/transaction.rs:
```diff
@@ -473,7 +547,22 @@ where
+                                    .ready()
+                                    .await
+                                    .expect("mempool poll_ready() method should not return an error")
+                                    .call(mempool::Request::CheckForVerifiedTransactions)
```
Why do we need to do this call here?

arya2 replied:
The docs for that variant say there are issues with polling a tower service with a `Buffer` layer without following that with a call.

upbqdn reviewed (dismissed):
What do you mean by "orphaned" transactions in the title?

upbqdn commented on zebra-consensus/src/transaction.rs:
```diff
@@ -570,7 +676,41 @@ where
         }
-        Ok((spent_utxos, spent_outputs))
+
+        if let Some(mempool) = mempool {
```
Edit: disregard this comment.

This new implementation changes the lookup logic. 

In the old implementation, if `is_mempool` was set, and `UnspentBestChainUtxo` returned `None`, the fn returned `TransactionError::TransparentInputNotFound`.

In the new implementation, if the same conditions hold, and `mempool` is `None`, no error is returned.

upbqdn replied:
Oh no, the above is wrong, closing.

arya2 committed:
Merge remote-tracking branch 'origin/main' into verify-orphaned-mempool-txs

arya2 committed:
Updates `spent_utxos()` method documentation

arya2 commented:
> What do you mean by "orphaned" transactions in the title?

Transactions that spend outputs that are missing in the chain state, though I'm not sure if "orphan transactions" is the right term if those outputs can still be found in the mempool.

This PR is still missing an acceptance test for the use case we're trying to address (testing it manually with zcashd seems difficult, and a test that creates transactions could provide a valuable reference for other tests in the future), but I think it's fine to merge it without an acceptance test since each component has unit tests (though I'm happy to prioritize writing that test if anyone disagrees).

upbqdn commented on zebra-consensus/src/transaction.rs:
```diff
@@ -514,28 +608,38 @@ where
+    /// the state for [`Block`](Request::Block) requests.
     ///
     /// `known_utxos` are additional UTXOs known at the time of validation (i.e.
     /// from previous transactions in the block).
```
###### Optional

```suggestion
```

upbqdn commented on zebra-consensus/src/transaction.rs:
```diff
@@ -514,28 +608,38 @@ where
         TransactionError,
     > {
+        let is_mempool = req.is_mempool();
+        let known_utxos = req.known_utxos();
```
###### Optional

```suggestion
        // Additional UTXOs known at the time of validation,
        // i.e., from previous transactions in the block.
        let known_utxos = req.known_utxos();
```

upbqdn commented on zebra-consensus/src/transaction.rs:
```diff
@@ -514,28 +608,38 @@ where
+    /// for them to be added to the mempool for [`Mempool`](Request::Mempool) requests.
+    ///
     /// Returns a tuple with a OutPoint -> Utxo map, and a vector of Outputs
     /// in the same order as the matching inputs in the transaction.
```
###### Optional

```suggestion
    /// Returns a triple containing:
    /// - `OutPoint` -> `Utxo` map, 
    /// - vec of `Output`s in the same order as the matching inputs in the `tx`,
    /// - vec of `Outpoint`s spent by a mempool `tx` that were not found in the best chain's utxo set.
```

upbqdn commented:
> > What do you mean by "orphaned" transactions in the title?

> Transactions that spend outputs that are missing in the chain state, though I'm not sure if "orphan transactions" is the right term if those outputs can still be found in the mempool.

This RFC, https://github.com/zcashfoundation/zebra/blob/257567b559dba0359519d61684e80daac52cc434/book/src/dev/rfcs/0005-state-updates.md#L62, defines orphaned blocks as those that are no longer included in the best chain, which implies they were mined at some point. I'd assume orphaned txs are those in such blocks, but this PR doesn't apply to them. Maybe something like "Verify transactions spending mempool UTXOs"?

upbqdn commented on zebrad/src/components/mempool/pending_outputs.rs:
```diff
@@ -0,0 +1,63 @@
+use zebra_node_services::mempool::Response;
+
+#[derive(Debug, Default)]
+pub struct PendingOutputs(HashMap<transparent::OutPoint, broadcast::Sender<transparent::Output>>);
```
Missing docs

arya2 replied:
Added in https://github.com/ZcashFoundation/zebra/pull/8857/commits/5043adf9e97e1e5e872ed77197b27e454f17efa5.

upbqdn commented on zebra-node-services/src/mempool/transaction_dependencies.rs:
```diff
@@ -0,0 +1,124 @@
+/// Representation of mempool transactions' dependencies on other transactions in the mempool.
+#[derive(Default, Debug, Clone)]
+pub struct TransactionDependencies {
+    /// Lists of mempool transactions that create UTXOs spent by
```
```suggestion
    /// Lists of mempool transaction ids that create UTXOs spent by
```

upbqdn commented on zebra-rpc/src/methods/get_block_template_rpcs/types/get_block_template.rs:
```diff
@@ -238,27 +243,38 @@ impl GetBlockTemplate {
             // Sort by hash, this is faster.
-            mempool_txs_with_templates
-                .sort_by_key(|(tx_template, _tx)| tx_template.hash.bytes_in_display_order());
+            mempool_txs_with_templates.sort_by_key(|(min_tx_index, tx_template, _tx)| {
```
Do we need to sort at all?

arya2 replied:
Probably not, I think it was there to make comparisons with the zcashd getblocktemplate method easier. I wouldn't mind removing it.

upbqdn replied:
I'm in favor of removing it as well then.

arya2 replied:
I've limited it to sorting only in test compilations in https://github.com/ZcashFoundation/zebra/pull/8857/commits/42dfc394501fa2a48dc2243682d94fadb37720b1 (it's still needed for the snapshot tests but I'd prefer to keep it out of release compilations).

upbqdn commented on zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs:
```diff
@@ -158,6 +178,52 @@ fn setup_fee_weighted_index(transactions: &[VerifiedUnminedTx]) -> Option<Weight
+    let Some(deps) = candidate_tx_deps else {
+        return true;
+    };
+
```
###### Optional

```suggestion

if selected_txs.len() < deps.len() {
    return false;
}

```

upbqdn commented on zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs:
```diff
@@ -181,30 +251,110 @@ fn checked_add_transaction_weighted_random(
+        let mut next_level_dependents = HashSet::new();
+
+        for dependent_tx_id in &current_level_dependents {
+            if has_direct_dependencies(tx_dependencies.get(dependent_tx_id), selected_txs) {
```
```suggestion
            // ## Note
            //
            // A necessary condition for adding the dependent tx is that it spends unmined outputs coming only from
            // the selected txs, which come from the mempool. If the tx also spends in-chain outputs, it won't
            // be added. This behavior is not specified by consensus rules and can be changed at any time,
            // meaning that such txs could be added.
            if has_direct_dependencies(tx_dependencies.get(dependent_tx_id), selected_txs) {
```

upbqdn commented on zebrad/src/components/mempool/tests/vector.rs:
```diff
@@ -956,6 +1083,8 @@ async fn setup(
         chain_tip_change.clone(),
     );
 
+    tokio::spawn(async move { while mempool_transaction_receiver.recv().await.is_ok() {} });
```
Why do we need this?

arya2 replied:
Dropping the receiver at the end of the `setup()` function could cause the mempool's `poll_ready()` method to return an error when trying to send transaction ids to a closed channel.

arya2 committed:
Avoids sorting getblocktemplate transactions in non-test compilations

arya2 committed:
documents PendingOutputs struct

arya2 committed:
Apply suggestions from code review

Co-authored-by: Marek <mail@marek.onl>

arya2 committed:
cargo fmt

upbqdn commented on zebra-rpc/src/methods/get_block_template_rpcs/types/get_block_template.rs:
```diff
@@ -237,28 +243,45 @@ impl GetBlockTemplate {
+                .into_iter()
+                .map(|(min_tx_index, tx)| (min_tx_index, (&tx).into(), tx))
+                .collect();
+            #[cfg(test)]
```
```suggestion
```

arya2 replied:
Thank you, and nice catch! Removed in https://github.com/ZcashFoundation/zebra/pull/8857/commits/df9c1c1894885e16fb02194130f8dca40097074d.

upbqdn commented on zebrad/src/components/mempool/storage/verified_set.rs:
```diff
@@ -161,49 +198,51 @@ impl VerifiedSet {
     /// [ZIP-401]: https://zips.z.cash/zip-0401
     #[allow(clippy::unwrap_in_result)]
-    pub fn evict_one(&mut self) -> Option<VerifiedUnminedTx> {
+    pub fn evict_one(&mut self) -> Vec<VerifiedUnminedTx> {
```
Returning a vec doesn't semantically correspond to the name of the fn and its docs.

upbqdn replied:
https://discord.com/channels/809218587167293450/809251029673312267/1304517119823712349

arya2 replied:
Fixed in https://github.com/ZcashFoundation/zebra/pull/8857/commits/df9c1c1894885e16fb02194130f8dca40097074d.

upbqdn commented on zebrad/src/components/mempool/storage/verified_set.rs:
```diff
@@ -161,49 +198,51 @@ impl VerifiedSet {
     ///
```
###### Optional

```suggestion
    ///
    /// This fn is equivalent to `EvictTransaction` in [ZIP-401].
    ///
```

arya2 replied:
Applied in https://github.com/ZcashFoundation/zebra/pull/8857/commits/df9c1c1894885e16fb02194130f8dca40097074d.

upbqdn commented on zebrad/src/components/mempool/storage.rs:
```diff
@@ -243,22 +270,25 @@ impl Storage {
-            // (we could return here, but we still want to check the mempool size)
-            if victim_tx.transaction.id == tx_id {
-                result = Err(SameEffectsChainRejectionError::RandomlyEvicted.into());
+            for victim_tx in victim_txs {
```
Evicting multiple txs at once doesn't seem to correspond to ZIP-401.

upbqdn replied:
Str4d made additional clarification: https://discord.com/channels/809218587167293450/809251029673312267/1304534822999556176.

arya2 replied:
> Str4d made additional clarification: https://discord.com/channels/809218587167293450/809251029673312267/1304534822999556176.

There doesn't seem to be any issue with adding rejection errors for all of the dependents, but it also doesn't need to allocate rejections for all of the dependents or complicate the code unnecessarily. ~I'll push a fix soon.~

Fixed in https://github.com/ZcashFoundation/zebra/pull/8857/commits/df9c1c1894885e16fb02194130f8dca40097074d.

upbqdn commented on zebrad/src/components/mempool/storage.rs:
```diff
@@ -605,31 +648,31 @@ impl Storage {
         // we need a separate set, since reject() takes the original unmined ID,
         // then extracts the mined ID out of it
-        let mut unmined_id_set = HashSet::new();
+        let mut mined_tx_ids = HashSet::new();
```
Let's document somewhere why we're using "mined" tx ids.

arya2 replied:
We actually don't need to use the unmined transaction ids here. We're using just the transaction hashes elsewhere in this PR since transactions that spend the outputs of other transactions in the mempool don't have the auth digest of those transactions in their transparent inputs, only the `OutPoint` (tx hash and output index).

Updated to avoid using the extra collection and with a comment on why mined ids are sufficient in  https://github.com/ZcashFoundation/zebra/pull/8857/commits/874372a8bfeabf08ce9aa903e62dfff8b97a566d.

upbqdn commented on zebrad/src/components/mempool/storage/verified_set.rs:
```diff
@@ -23,14 +26,26 @@ use zebra_chain::transaction::MEMPOOL_TRANSACTION_COST_THRESHOLD;
+    /// The [`transparent::Utxo`]s created by verified transactions in the mempool
+    ///
+    /// Note that these UTXOs may not be unspent.
+    /// Outputs can be spent by later transactions or blocks in the chain.
```
Shouldn't we go with outputs instead of UTXOs?

```suggestion
    /// The [`transparent::Output`]s created by verified transactions in the mempool
    ///
    /// Note that these outputs may not be unspent.
    /// They can be spent by later transactions or blocks in the chain.
```

arya2 replied:
Yes we should. Applied in https://github.com/ZcashFoundation/zebra/pull/8857/commits/874372a8bfeabf08ce9aa903e62dfff8b97a566d.

upbqdn commented on zebrad/src/components/mempool/storage/verified_set.rs:
```diff
@@ -23,14 +26,26 @@ use zebra_chain::transaction::MEMPOOL_TRANSACTION_COST_THRESHOLD;
+    /// The [`transparent::Utxo`]s created by verified transactions in the mempool
+    ///
+    /// Note that these UTXOs may not be unspent.
+    /// Outputs can be spent by later transactions or blocks in the chain.
```
How can they be spent by transactions in the chain?

arya2 replied:
They can't, the documentation here was wrong, thank you for catching that. Fixed in https://github.com/ZcashFoundation/zebra/pull/8857/commits/874372a8bfeabf08ce9aa903e62dfff8b97a566d.

upbqdn commented on zebrad/src/components/mempool/storage/verified_set.rs:
```diff
@@ -241,8 +288,19 @@ impl VerifiedSet {
+    /// Inserts the transaction's outputs into the internal caches and responds to pending output requests.
+    fn cache_outputs_and_respond_to_pending_output_requests_from(
+        &mut self,
+        tx_hash: transaction::Hash,
```
###### Optional

We're passing `tx_hash` separately even though it's part of `tx`. Also, we call this fn only in one spot, and the body is trivial, so we could just remove the fn altogether.

arya2 replied:
> so we could just remove the fn altogether

Ah, thank you! The updated method name was far too verbose. Removed in https://github.com/ZcashFoundation/zebra/pull/8857/commits/874372a8bfeabf08ce9aa903e62dfff8b97a566d

upbqdn reviewed (commented):
I'm trying to figure out what happens if someone submits txs containing a circular dependency. We should make sure we can handle such cases well.

arya2 commented:
> I'm trying to figure out what happens if someone submits txs containing a circular dependency. We should make sure we can handle such cases well.

The transaction verifier would return an `TransparentInputNotFound` error for both because (unlike pending_utxos in the state service) the mempool only responds to `AwaitOutput` requests once the transaction creating that output has been added to the mempool's verified set.

A similar question around ordering the transactions in block templates caused me some strife (the code should work as expected there too, and it's been excluded from release compilations, but it was difficult to reason about). That exact question may have also caused me some strife a little earlier, as well as what happens to long chains of dependent transactions.

arya2 committed:
Applies suggestions from code review

Avoids unnecessarily rejecting dependent transactions of randomly evicted mempool transactions.

Updates `TransactionDependencies::remove_all()` to omit provided transaction id from the list of removed transaction ids.

arya2 committed:
Applies suggestions from code review.

arya2 committed:
Adds minor comments

upbqdn commented on zebrad/src/components/mempool/storage/verified_set.rs:
```diff
@@ -23,14 +26,25 @@ use zebra_chain::transaction::MEMPOOL_TRANSACTION_COST_THRESHOLD;
 /// outputs include:
 ///
+/// - the dependencies of transactions that spent the outputs of other transactions in the mempool
+/// - the UTXOs of transactions in the mempool
```
Shouldn't this be outputs as well?


```suggestion
/// - the outputs of transactions in the mempool
```

arya2 committed:
Update zebrad/src/components/mempool/storage/verified_set.rs

Co-authored-by: Marek <mail@marek.onl>

upbqdn committed:
Remove an outdated comment (#9013)

gustavovalverde commented:
@mergify refresh

gustavovalverde commented:
@mergify requeue

gustavovalverde commented:
@mergify queue
